Adding WITH ORDINALITY to DuckDB

niykko opened the pull request:
_This is a continuation of #9014_

I am resuming work on this function. If I understand correctly, you still want me to implement this feature by internally producing a **_ROW_NUMBER()_** window function instead of producing the column manually.

My first step is to take a look again at how these operators work and then try to find a solution.

niykko committed:
Reviving the project: activate in parser, add request type

niykko committed:
Remove automatically generated include (no idea how this happened, starting to lose my mind over here)

Mytherin commented:
Thanks for the PR!

I think there are two distinct steps that need to happen to have `WITH ORDINALITY` work:

* The scanner table functions (e.g. `read_parquet`, etc) that do not support lateral joins can support `WITH ORDINALITY` by pushing a `ROW_NUMBER() OVER ()`
* The table in-out functions that do support lateral joins need to explicitly handle `WITH ORDINALITY`, and compute the ordinality column during execution (ideally in `PhysicalTableInOutFunction`).

Ideally this is generated during the bind phase already - i.e. when binding a scanner table function that has ordinality specified, the `LogicalWindow` with `row_number() over ()` is generated. The optimizer can then (potentially) remove this again or do something more clever with this.

niykko committed:
Next steps

niykko committed:
seperate runtime and statistical data fields

niykko committed:
fixes: namespace, comparison operators, name refactor

niykko committed:
explicit cast instead of implicit

niykko committed:
oversight, wrong index fixed

niykko committed:
basic functionality for InOut functions, now need to handle resets

niykko committed:
works laterally for InOut Functions

niykko committed:
Added LogicalWindow

niykko committed:
run make format-fix (this also touched some .py files i have no business with, i discarded these changes)

niykko commented:
I'd like to note that when using any version higher than **24.1.0** for the python module **black**, the following files get changed when running _"make format-fix"_:

> modified:   scripts/regression/benchmark.py
>         modified:   tools/pythonpkg/tests/fast/numpy/test_numpy_new_path.py
>         modified:   tools/pythonpkg/tests/fast/udf/test_null_filtering.py
>         modified:   tools/pythonpkg/tests/fast/udf/test_scalar.py

Just so you know, it seems like this package has changed the formatting rules. It now inserts blank lines after certain comments.

niykko committed:
renamed ordinality_data type to ordinality_data_t

niykko committed:
add SetOrdinality when input is not projected

niykko committed:
refactor: removed unneeded reset bool

niykko committed:
removed corresponding entry in sourcestates

niykko committed:
manually added serialization for TableFunctionRef

niykko committed:
add proper serialization (had to find the right file first)

niykko committed:
ran serialization and enum scripts

niykko committed:
remove previously added json entry, add test

niykko committed:
add WITH ORDINALITY to ToString() of tablefunctioref

niykko committed:
actually return result ;)

niykko committed:
changed some tests

niykko commented:
Right now there are two problems left.

- When using _read\_*_ functions, if the column names aren't specified and any entry within the first row is called "ordinality", we get an error because the **WITH ORDINALITY** column is already called "ordinality". In order to fix this, I would need to find out at which point in the MultiFileReader process I can get access to the read column names. I would then add a casted integer value to the WITH-ORDINALITY-column-name-string until the column name can not be found in the column names found in the csv file.
- I am missing something with the Serialization, but can not figure out what it is. The original queries work, but with the deserialized queries in the unit test, something goes wrong 🤔

niykko committed:
fix output string

niykko committed:
fix serialization

niykko committed:
fixed some unittests

niykko committed:
Merge branch 'main' into with_ordinality_reloaded

niykko committed:
refactor: remove index in globalStates. needs to be added back in in the future if the calls to the tablefunctions  were to change again: we need to make sure to reset the index ffor new sources. not needed anymore right now (and probably wont be in the future, was a leftover from old PR)

niykko committed:
continue ordinality over chunk borders

niykko committed:
moved reset operation to new row instead of new chunk

niykko committed:
run format-fix

niykko commented on extension/json/include/json_enums.hpp:
These changes were made because of the serialization script.

niykko commented:
I am currently having trouble with the serialization & deserialization, as the "Storage Size regression" test fails. 
The key information seems to be:
`Regenerate test/sql/storage_version.db with newer version -> read with older version`
and
`Error: unable to open database "test/sql/storage_version/storage_version.db": Serialization Error: Failed to deserialize: expected end of object, but found field id: 202`

The **field id** this is referring to, 202, is the id that I gave the `ordinality_idx` that is now part of a TableRef. 

It seems to me like something is serialized with my newer version of DuckDB and then read by an older version.

How can I fix this?

niykko committed:
Merge branch 'main' into with_ordinality_reloaded

niykko committed:
follow instructions given in storage_version.test_slow, but change version_map.json this time instead of manually

niykko committed:
change include

niykko committed:
block parallel execution for InOutFunctions when WITH ORDINALITY is specified + small test change

niykko committed:
include types explicitly

niykko committed:
include logicalprojection in bind_table_function (locally worked because of namespaces but get error in vcxproj)

niykko committed:
refactor name of Ordinality_request_t, reset storage version, rerun scripts

niykko committed:
rerun storage script, now back to original

niykko committed:
version nr

niykko committed:
switch critical lines in binding

niykko committed:
merge upstream

niykko committed:
move bind line

niykko committed:
transfer ordinality information from logical plan to table func

niykko committed:
add explicit default for tableref

niykko committed:
Merge remote-tracking branch 'upstream/main' into with_ordinality_reloaded

niykko committed:
update serialized_plans.binary

niykko commented:
The unittest 'Test deserialized plans from file' is failing with a column bind error.
I was trying to hunt down the origin of this error for a while, but then noticed that this failure occurs on the production branch of duckdb as well. 

Steps to replicate:

call unittest on the debug version and test 'Test deserialized plans from file'

niykko committed:
format-fix

niykko committed:
added tests

niykko committed:
not ready for review, inserted wrong tests

niykko committed:
delegate table-function ordinality entirely to window operator (logical get is not informed anymore & correct the tests (typos)

niykko committed:
format-fix

niykko committed:
remove OS specific glob test

niykko committed:
safe index conversion

niykko committed:
update test description & move csv files

niykko committed:
change static cast to numeric cast

niykko committed:
Merge branch 'main' into with_ordinality_reloaded

niykko commented:
Hey @Mytherin ,
this PR is now ready for review. Please let me know what you think 😃 
Please also note that the failing unittest ('Test deserialized plans from file') also fails on the production branch, so I assume my additions are not what causes this error.

niykko committed:
update indexes & remove #include from merge

niykko committed:
update storage file (run 'Generate serialized plans file')

niykko committed:
Merge remote-tracking branch 'upstream/main' into with_ordinality_reloaded

niykko committed:
order matters when deserializing: move line according to deserialization correct order

niykko committed:
Merge remote-tracking branch 'upstream/main' into with_ordinality_reloaded

niykko committed:
use mutex for changing globaloperator ordinality data

niykko committed:
Merge remote-tracking branch 'upstream/main' into with_ordinality_reloaded

Mytherin commented on src/execution/operator/projection/physical_tableinout_function.cpp:
```diff
@@ -87,6 +96,8 @@ OperatorResultType PhysicalTableInOutFunction::Execute(ExecutionContext &context
 		state.row_index++;
 		state.new_row = false;
+		lock_guard<mutex> guard(gstate.ordinality_lock);
+		gstate.current_ordinality_idx = 1;
```
Given that the other scan state is all located in the local state - I suspect the global state is the wrong place for the `current_ordinality_idx`. Do we have a test with a larger `UNNEST` or similar table function that triggers multi-threading?

Mytherin commented on src/execution/operator/projection/physical_tableinout_function.cpp:
```diff
@@ -68,7 +70,14 @@ OperatorResultType PhysicalTableInOutFunction::Execute(ExecutionContext &context
+		auto result = function.in_out_function(context, data, input, chunk);
+		if (function.ordinality_data.ordinality_request == Ordinality_request_t::REQUESTED) {
+			const idx_t ordinality = chunk.size();
+			function.ordinality_data.SetOrdinality(chunk, gstate.current_ordinality_idx, ordinality);
```
We are reading `gstate.current_ordinality_idx` without a lock, whereas we are modifying it below with a lock - one of these is wrong. I suspect we need to move this to the thread-local state anyway however, in which case we don't need a lock at all.

Mytherin commented on src/execution/operator/scan/physical_table_scan.cpp:
```diff
@@ -120,6 +123,12 @@ SourceResultType PhysicalTableScan::GetData(ExecutionContext &context, DataChunk
 
+	if (function.ordinality_data.ordinality_request == Ordinality_request_t::REQUESTED) {
+		idx_t ordinality = chunk.size();
+		function.ordinality_data.SetOrdinality(chunk, g_state.ordinality_idx, ordinality);
```
For `PhysicalTableScan` - I think we can instead perform this using a window function. This will currently be incorrect when multi-threading.

Mytherin commented on src/include/duckdb/function/table_function.hpp:
```diff
@@ -420,6 +421,8 @@ class TableFunction : public SimpleNamedParameterFunction { // NOLINT: work-arou
 	//! Whether or not the table function supports late materialization
 	bool late_materialization;
+	//! Data for WITH ORDINALITY
+	ordinality_data_t ordinality_data;
```
The table function is the wrong place for this - the TableFunction should not contain bind information. This should be stored in the bound nodes (`TableFunctionRef`/`LogicalGet`/...) instead

Mytherin commented on test/sql/ordinality/a0.csv:
```diff
@@ -0,0 +1,11 @@
+Numbers, Chars
```
Instead of adding all of these data files - can we use existing data files or generate data in the tests themselves?

Mytherin commented on test/sql/ordinality/ordinality_readcsv.test_slow:
```diff
@@ -0,0 +1,18 @@
+query III
+SELECT * FROM read_csv('test/sql/ordinality/a1.csv') WITH ORDINALITY ORDER BY Numbers,Chars,ordinality;
+----
+7800 values hashing to d86e8a5e8df33cec28e0ed7b6c386419
```
In general we prefer to avoid using the hash comparison feature - we can verify two results are the same using the labeled results, we can use that to verify results instead, e.g.:

```sql
query III nosort read_csv_result
SELECT * FROM read_csv('test/sql/ordinality/a1.csv') WITH ORDINALITY ORDER BY Numbers,Chars,ordinality;

query III nosort read_csv_result
SELECT *, row_number() OVER () AS ordinality FROM read_csv('test/sql/ordinality/a1.csv') ORDER BY Numbers,Chars,ordinality;
```

Mytherin reviewed (commented):
Thanks for the PR! Great that this is being picked up again. Looks good - some comments below.

It would be great to get some more tests going as well:

* WITH ORDINALITY on larger tables, e.g. you could use the TPC-H data generator and then create tests with those
* WITH ORDINALITY used in a view definition
* WITH ORDINALITY **on** a view (should this even work?)
* WITH ORDINALITY on a CTE
* WITH ORDINALITY on a recursive CTE
* WITH ORDINALITY used within a recursive CTE
* WITH ORDINALITY where the table already contains a column called "ordinality"
* Filters on the ORDINALITY column
* Projecting out all columns except for the ORDINALITY column

niykko committed:
restore serialized_plans.binary

niykko committed:
revert storage_version.db

niykko committed:
rename ordinality types

niykko committed:
run enum_util script

niykko committed:
further refactor type names (not caught by initial refactor)

niykko committed:
move state information from local to global state

niykko committed:
move bind information from table function to LogicalGet and then propagate to physical scan

niykko committed:
update plan creation for non-correlated InOut

niykko committed:
change tests (first phase, not finished) and remove serialization entirely for a clean start. take care of serialization next

niykko committed:
run make generate-files

niykko committed:
re-add serialiation only for the request (does not work yet...)

niykko committed:
proper serialization

niykko committed:
do not add WITH_ORDINALITY request after serialization when InOut function is not correlated (window function)

niykko committed:
make format-main and additional tests

niykko committed:
replace * in select with explicit aliases

niykko committed:
seperate tests for InOut and constant TVFs + additional tests

niykko committed:
remove deprecated testfiles

niykko committed:
move unnest-tests into appropriate new test files

niykko committed:
for InOut-functions on constants push window function (like for regular TVFs) and only get physical for correlations

niykko committed:
make format-main

niykko committed:
change tests

niykko committed:
removed old debug-line

niykko committed:
move view test to constant tests and add unnest test

niykko committed:
add test & add forgotten return-value for other test

niykko committed:
run make-format main & fix typo in test group

niykko committed:
explicit safe cast + dynamic ordinality column naming

niykko commented:
I have made the appropriate changes with regards to your comments (very useful, I like the code much better now, thanks) 👍 

However, there are still some points that I would like to give my thoughts on and get some clarification: 

>     * WITH ORDINALITY **on** a view (should this even work?)

I do not think that **WITH ORDNALITY** is particularly useful here, my reasoning being that any order forced onto the view using the feature could instead be pushed down and made part of the view instead. 

>     * WITH ORDINALITY on a CTE
>     * WITH ORDINALITY on a recursive CTE

I would argue the same for these cases. Additionally, I would have to tweak the parser. The SQL-standard also only defines **WITH ORDINALITY** for TVFs.

>     * WITH ORDINALITY where the table already contains a column called "ordinality"

As Postgres does not support **WITH ORDINALITY** for .csv files since there is no TVF like **read_csv** _(big bummer! That's where this feature shines the most in my opinion)_ I could not copy its behaviour but had to think of something myself.
So now for this case, the columns are dynamically named by iterating a suffix until the name is free. For example, if my table (from a CSV file or otherwise, this also holds for InOut TVFs) contains the column _'ordinality'_ then the **WITH ORDINALITY**-column will be named _'ordinality0'._ If the colum 'ordinality0' also exists already, it will be called 'ordinality1' and so on. Please let me know if you prefer a different naming scheme or different behaviour. 

Thank you for your time and effort reviewing this PR🙏

niykko committed:
Merge remote-tracking branch 'upstream/main' into with_ordinality_reloaded

niykko committed:
serialization error is back, write with default instead

Mytherin reviewed (commented):
Thanks! Looking better - some more comments

niykko committed:
remove leftover fields

niykko committed:
remove inout_correlation as its unnecessary now

niykko committed:
replace OrdinalityData struct with optional_idx and be a bit smarter about the logic

niykko commented on src/planner/binder/tableref/plan_table_function.cpp:
```diff
@@ -6,7 +6,21 @@ namespace duckdb {
+
+		LogicalOperator *node = ref.get.get();
+
+		while (!node->children.empty()) {
```
Before now, this function assumes that _ref_ is a _LOGICAL_GET_ and adds the subquery-plan as a child of this node.

When we push a window function, however, we place this function above the LOGICAL_GET (as well as a projection to extract the values). _ref_ then look like this:

        PROJECTION
            |
          WINDOW
            |
       LOGICAL_GET
       
This would result in us placing the subquery-plan as a second child of the projection, when we really want the output of this plan to feed the _LOGICAL_GET_.

But instead of statically inserting the subquery plan at the second child, I dynamically look for the last 'node' within this partial plan. This is just in case in the future (or right now and I just don't know about it), it may become a possibility that more or less than exactly two nodes are added when pushing a window function. I would also have to carry additional information about whether or not a window function has been pushed, which I avoid by doing it this way. When this is not the case, all we 'lose' is a pointer copy, which is why I think this solution is sensible.

Please let me know what you think of this and whether you prefer a static (or different) approach instead.

niykko replied:
This picture demonstrates what the full plan for 

`select * from range((select 5)) WITH ORDINALITY;`


looks like afterwards:

![image](https://github.com/user-attachments/assets/3dd4add9-f77f-4d7f-ab72-fe2cd0712774)

niykko committed:
Merge branch 'duckdb:main' into with_ordinality_reloaded

niykko committed:
Merge remote-tracking branch 'upstream/main' into with_ordinality_reloaded

Mytherin reviewed (commented):
Thanks for the changes! Looks good - some minor nits then this is good to go

niykko committed:
move SetOrdinality function from enum header to PhysicalTableInOut && remove includes

niykko committed:
consider aliases as well when deciding on the final column name for with ordinality

niykko committed:
change raw pointer to reference

niykko committed:
format-fix

niykko commented:
I moved SetOrdinality from the enum header into PhysicalTableInOut, removed unneccessary includes and changed the pointer to a reference.

Please let me know your thoughts.

niykko committed:
change alias test since aliases are now considered as well

niykko committed:
re-introduce types include to fix tidy check

soerenwolfers commented:
@niykko Any reason `WITH ORDINALITY` isn't allowed on regular table references? That'd be a very frequent use case for me.

niykko commented:
@soerenwolfers 

I think WITH ORDINALITY is most useful when importing data from a sequential container into relational system when this data has an order that is not explicitly encoded within the data itself but instead given by the position of the items. At the point of import (for example by using _**read_csv**_ to read from a .csv file or _**unnest**_ to read from an array, both being linear data structures) this order is then copied and enforced by WITH ORDINALITY.

When you are working with an already constructed regular table and invoke WITH ORDINALITY, since it is a relational object, there is no order to be translated unless it is already encoded within the table itself. And in that case, WITH ORDINALITY would just be syntactic sugar for expressions like

`SELECT ROW_NUMBER() OVER () AS ordinality, t.* FROM table AS t `

soerenwolfers commented:
@niykko Duckdb does already provide order guarantees for its tables (see https://duckdb.org/docs/stable/sql/dialect/order_preservation.html) beyond the pure relational model. 

As such, `WITH ORDINALITY` is well defined in duckdb -- in fact, as you say, it should return the exact same column that `row_number() OVER ()` returns but with less typing.  Maybe I'm missing something, but my impression was that the same applied to everything in this PR, i.e., that you could already achieve it before, just with more typing. 

Also, note that for many users, table references are actually replacement scanned for, e.g., pandas data frames, and those definitely have a well defined order.
